### PR TITLE
ref: clean config [1/n] add intermediate setters

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

preparation step for cleaning the distributed and device logic...

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM ! Happy to finally see this part being cleaned out !

**Review by SeanNaren — APPROVED** [review]

**Comment by Borda** [comment]
the failing static type checking was introduced in another PR...

**Comment by Borda** [comment]
@williamFalcon mind review?
note it does not change any logic nor calls, this shall be done in followup PR

File: pytorch_lightning/utilities/__init__.py
Comment (carmocca) [inline pytorch_lightning/utilities/__init__.py:65]:
```suggestion
class DistributedType(str, Enum):
```

If you make string enums inherit from str, you can do:

```
DistributedType.DP == "dp"
```

which is very convenient to avoid casting from/to str. Same for the other string enums in this file

File: pytorch_lightning/trainer/deprecated_api.py
Comment (awaelchli) [inline pytorch_lightning/trainer/deprecated_api.py:45]: If you add these, the user will see many of the warnings because internally the setter is used.
Is the deprecation really necessary? The user is not supposed to interact with these attributes in the first place, right?

Comment (Borda) [reply pytorch_lightning/trainer/deprecated_api.py:45]: ok, I ll drop the warning for now and add it after we replace it in code...

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
@Borda we should decide whether we want to go with uppercase naming or not and be consistent with it.

**Comment by Borda** [comment]
> @Borda we should decide whether we want to go with uppercase naming or not and be consistent with it.

ok, I would then go with the upper case as it is easy to convert (compare to mix like `DDP_spawn`) and apply with most cases

File: pytorch_lightning/utilities/__init__.py
Comment (carmocca) [inline pytorch_lightning/utilities/__init__.py:61]: Can we use a better name? `StrEnum` or `LightningEnum`

fun fact: `StrEnum` will be part of the sdtlib in Python 3.10 (https://bugs.python.org/issue41816) (backport: https://github.com/clbarnes/backports.strenum)

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py
Comment (carmocca) [inline tests/trainer/test_trainer.py:1345]: Does this solve your second TODO?

```suggestion
    for k, v in expected.items():
        assert getattr(trainer, k) == v
```

File: tests/trainer/test_trainer.py
Comment (carmocca) [inline tests/trainer/test_trainer.py:1336]: This should fix these TODOs:

```python
@pytest.mark.parametrize(
    "trainer_kwargs,expected",
    [
        (
            dict(distributed_backend=None, gpus=None),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="dp", gpus=None),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="dp", gpus=None),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp", gpus=None),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp", num_processes=2, gpus=None),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=2,
            ),
        ),
        (
            dict(distributed_backend="ddp", num_nodes=2, gpus=None),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp_cpu", num_processes=2, gpus=None),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=2,
            ),
        ),
        (
            dict(distributed_backend="ddp2", gpus=None),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend=None, gpus=1),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=1,
                on_gpu=True,
                use_single_gpu=True,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="dp", gpus=1),
            dict(
                use_dp=True,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=1,
                on_gpu=True,
                use_single_gpu=True,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp", gpus=1),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=1,
                on_gpu=True,
                use_single_gpu=True,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp_cpu", num_processes=2, gpus=1),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=0,
                on_gpu=False,
                use_single_gpu=False,
                num_processes=2,
            ),
        ),
        (
            dict(distributed_backend="ddp2", gpus=1),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=True,
                num_gpus=1,
                on_gpu=True,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend=None, gpus=2),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=2,
                on_gpu=True,
                use_single_gpu=False,
                num_processes=2,
            ),
        ),
        (
            dict(distributed_backend="dp", gpus=2),
            dict(
                use_dp=True,
                use_ddp=False,
                use_ddp2=False,
                num_gpus=2,
                on_gpu=True,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
        (
            dict(distributed_backend="ddp", gpus=2),
            dict(
                use_dp=False,
                use_ddp=True,
                use_ddp2=False,
                num_gpus=2,
                on_gpu=True,
                use_single_gpu=False,
                num_processes=2,
            ),
        ),
        (
            dict(distributed_backend="ddp2", gpus=2),
            dict(
                use_dp=False,
                use_ddp=False,
                use_ddp2=True,
                num_gpus=2,
                on_gpu=True,
                use_single_gpu=False,
                num_processes=1,
            ),
        ),
    ],
)
def test_trainer_config(trainer_kwargs, expected, monkeypatch):
    if trainer_kwargs["gpus"] is not None:
        monkeypatch.setattr(torch.cuda, "is_available", lambda: True)
        monkeypatch.setattr(torch.cuda, "device_count", lambda: trainer_kwargs["gpus"])
    trainer = Trainer(**trainer_kwargs)
    for k, v in expected.items():
        assert getattr(trainer, k) == v, f"Failed {k}: {v}"
```

Comment (Borda) [reply tests/trainer/test_trainer.py:1336]: nice! I would clean up the tests a bit more but for clarity that the functionality is not touched I would rather keep the tests as they are and you are welcome to send a PR just with this tests update :]

Comment (carmocca) [reply tests/trainer/test_trainer.py:1336]: Sure! I'll open a PR with this patch once this PR is merged

File: pytorch_lightning/accelerators/accelerator.py
Comment (williamFalcon) [inline pytorch_lightning/accelerators/accelerator.py:258]: can we avoid enums please?
it's always impossible to figure out how to import them etc...

**Review by williamFalcon — CHANGES_REQUESTED** [review]

File: pytorch_lightning/accelerators/accelerator.py
Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:258]: Id they inherit `str`, then they are optional for users. And we can still use the enum internally for nicer code (my opinion)

Comment (Borda) [reply pytorch_lightning/accelerators/accelerator.py:258]: as they inherit from str, you can compare with string so resolving the issue you faced last time.. see the examples

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:23]: Can you use LightningEnum everywhere ?

**Pull request merged**
